### PR TITLE
Add card advanced filter for Search

### PR DESCRIPTION
**Pull request by 289Adam289**
### Fixed Issues

$[46042](https://github.com/Expensify/App/issues/46042)
PROPOSAL:

### Tests

- add cards to account
- visit route `/search/filters`
- test card filter and whether navigation behaves correctly

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/dd8d4729-d02f-436a-be49-8fbac30ea56b
</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/435e3863-e369-4ef8-81df-a195dae6aabf

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/b3a9300d-f668-4a28-900d-b97f22e3c8f7
</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/18668a86-15d8-495d-9ed2-cb7a3acc9ed5

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/7093b727-aefe-4e62-a742-90c26d382137

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/096a7b93-3f21-46f9-8828-5cc5708ed5d3

</details>

### CODE REVIEW

File: src/pages/Search/SearchFiltersCardPage.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    const [cardList] = useOnyx(ONYXKEYS.CARD_LIST);
+    const [searchAdvancedFiltersForm] = useOnyx(ONYXKEYS.FORMS.SEARCH_ADVANCED_FILTERS_FORM);
+    const currentCards = searchAdvancedFiltersForm?.cardID;
+    const [newCards, setNewCards] = useState(currentCards ?? []);
+
+    const sections = useMemo(() => {
+        const newSections: CategorySection[] = [];
+        const cards = Object.values(cardList ?? {})
+            .filter((card) => card.cardID)
```
Comment (Kicu) [inline src/pages/Search/SearchFiltersCardPage.tsx:34]: Im not sure why we need this

File: src/pages/Search/SearchFiltersCardPage.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            .sort((a, b) => a.bank.localeCompare(b.bank))
+            .map((card) => {
+                const icon = getBankIcon({bankName: card.bank as BankName, isCard: true, styles});
```
Comment (Kicu) [inline src/pages/Search/SearchFiltersCardPage.tsx:37]: This cast `as BankName` looks fishy and I'd rather we didn't have it. But I also did a search in the project and it seems that it is done similarly in a few places so maybe thats ok.
Leaving this comment for other people to see

File: src/pages/Search/SearchFiltersCardPage.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    }, [cardList, styles, newCards]);
+
+    const updateCard = useCallback((values: Partial<FormOnyxValues<typeof ONYXKEYS.FORMS.SEARCH_ADVANCED_FILTERS_FORM>>) => {
+        SearchActions.updateAdvancedFilters(values);
+    }, []);
+
+    const handleConfirmSelection = useCallback(() => {
+        updateCard({
+            cardID: newCards,
+        });
```
Comment (Kicu) [inline src/pages/Search/SearchFiltersCardPage.tsx:61]: your `updateCard` function is just a oneliner, so it's really not needed in here.
Just call `updateAdvancedFilters()` directly in this line

File: src/pages/Search/SearchFiltersCardPage.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    const updateNewCards = useCallback(
+        (item: Partial<OptionData>) => {
+            if (!item.keyForList) {
+                return;
+            }
+            if (item.isSelected) {
+                setNewCards(newCards?.filter((card) => card !== item.keyForList));
+            } else {
+                setNewCards([...(newCards ?? []), item.keyForList]);
```
Comment (Kicu) [inline src/pages/Search/SearchFiltersCardPage.tsx:73]: you are defaulting `newCards` to `[]` so you don't need `?? []` in here

**Review by Kicu — APPROVED** [review]
Looks good, left you small comments

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by ikevin127** [comment]
@Kicu Navigation and selection of the cards is pristine 🚀 

⚠️ I completed the **PR Reviewer Checklist**, with the mention that I found possibly one / two potential issues during testing:

* [ ] **1. Missing design for when no cards are present (empty state)**

    **Note:** Not sure if we're supposed to even show the empty card filter page if cards don't exist as in this case we could not show the `Card` filtering category at all / or show it grayed-out `Card` filtering category with an ℹ️ `You don't have any cards added yet.` kind of tooltip / message.
    | Empty card filter RHP |
    | ------------- |
    | no-cards-empty|

##

* [ ] **2. Card filter labels are not added in the Filters section even though they are present in the URL**
    
    **Note:** Not sure if this is because I added the cards via Onyx or this happens regardless if the cards are actually added.
    | Missing Card filter labels |
    | ------------- |
    | Screenshot 2024-08-02 at 15 26 01 |

##

@luacmartins @Expensify/design If you guys think these are not within the immediate scope of this PR and will be handled in a follow-up / or when all the filters will be linked together then let me know and I'll 🟢 **Approve** right away.

**Comment by dubielzyk-expensify** [comment]
I think both of these needs to be fixed. I'm a bit unsure if we want an empty state for the card or if we wanna remove the filter option altogether when there are no cards. If we did the empty state route, we could do something like this:

image

Perhaps it's nice to have the filter option in there purely for education purposes, but I'm not sure it makes sense. cc @Expensify/design and @JmillsExpensify for thoughts

**Comment by luacmartins** [comment]
I agree that we need to fix both issues. I like the idea of simplifying the UI and only showing relevant filters to the user, so I'd be in favor of removing the card filter if the user doesn't have any cards.

**Comment by Kicu** [comment]
Are guys suggestion that we remove the very option of card filter from the main "advanced filters" page?

Screenshot 2024-08-05 at 10 14 59

**Comment by 289Adam289** [comment]
The issue with card labels in the filters section, I believe, is a result of a discrepancy between the key in the search parser grammar: "card", and the key in the `CONST.ts` file: "cardID". Changing the key in the grammar to "cardID" should fix the issue.

**Comment by luacmartins** [comment]
Remind the option from the advanced filter page if the user doesn't have any cards would be my preference, but we should wait on more input from the design team.

**Comment by 289Adam289** [comment]
The parser wasn't regenerated after the last grammar change, which is why the `searchParser.js` file has changed significantly. cc @Kicu @adamgrzybowski

File: src/ROUTES.ts [hunk 5]
```diff
@@ -55,6 +55,8 @@ const ROUTES = {
 
     SEARCH_ADVANCED_FILTERS_CATEGORY: 'search/filters/category',
 
+    SEARCH_ADVANCED_FILTERS_CARD: 'search/filters/card',
```
Comment (Kicu) [inline src/ROUTES.ts:58]: remove emptyline before this route, as Jakub is removing all the empty lines within search routes definitions.

**Review by Kicu — COMMENTED** [review]
I left just 1 tiny comment :D

**Comment by shawnborton** [comment]
I think I lean towards just removing the filter option entirely if there are no cards to filter on.

**Comment by dannymcclain** [comment]
>I think I lean towards just removing the filter option entirely if there are no cards to filter on.

I'm down with this approach.

**Review by ikevin127 — APPROVED** [review]
🟢 Thanks for addressing the issues!

##

* [x] **1. Missing design for when no cards are present (empty state)**

    Addressed by the team and decided to go with removing the `Card` category from the filters list when no cards are present. Fixed by commit https://github.com/Expensify/App/pull/46666/commits/1f000e44a51a46b031c6292c961feb3d0b982bc7.

##

* [x] **2. Card filter labels are not added in the Filters section even though they are present in the URL**
    
    Addressed by author and fixed by commit https://github.com/Expensify/App/pull/46666/commits/f26e0d923fb98617c81bc5f5d9b3d8bbb128ad72. This is how it looks now (comma in-between cards):
    Screenshot 2024-08-05 at 11 31 20

**Comment by ikevin127** [comment]
Regarding the commas in-between the **cardID**, this might be addressed in another PR as per https://github.com/Expensify/App/pull/46566#issuecomment-2268627463.
I wouldn't consider it a blocker for this PR.
@Kicu should be able to confirm if the fix mentioned in the comment will cover all filters (including `Card`).

**Review by luacmartins — APPROVED** [review]
LGTM

**Pull request merged**